pkg/apis/messaging/v1alpha1/pipeline_types.go

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright 2019 The Knative Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package v1alpha1
+
+import (
+	eventingv1alpha1 "github.com/knative/eventing/pkg/apis/eventing/v1alpha1"
+	"github.com/knative/pkg/apis"
+	duckv1alpha1 "github.com/knative/pkg/apis/duck/v1alpha1"
+	"github.com/knative/pkg/webhook"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+type Pipeline struct {
+	metav1.TypeMeta `json:",inline"`
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the desired state of the Pipeline.
+	Spec PipelineSpec `json:"spec,omitempty"`
+
+	// Status represents the current state of the Pipeline. This data may be out of
+	// date.
+	// +optional
+	Status PipelineStatus `json:"status,omitempty"`
+}
+
+// Check that Pipeline can be validated, can be defaulted, and has immutable fields.
+var _ apis.Validatable = (*Pipeline)(nil)
+var _ apis.Defaultable = (*Pipeline)(nil)
+var _ apis.Immutable = (*Pipeline)(nil)
+var _ runtime.Object = (*Pipeline)(nil)
+var _ webhook.GenericCRD = (*Pipeline)(nil)
+
+// This should be duck so that Broker can also use this
+type ChannelTemplateSpec struct {
+	// You can specify only the following fields of the ObjectReference:
+	//   - Kind
+	//   - APIVersion
+	ChannelCRD corev1.ObjectReference `json:"channelCRD,omitempty"`
+
+	// Spec defines the Spec to use for each channel created. Passed
+	// in verbatim to the Channel CRD as Spec section.
+	// +optional
+	Spec *runtime.RawExtension `json:"spec,omitempty"`
+}
+
+type PipelineSpec struct {
+	// Steps is the list of Subscribers (processors / functions) that will be called in the order
+	// provided.
+	Steps []eventingv1alpha1.SubscriberSpec `json:"steps"`
+
+	// ChannelTemplate specifies which Channel CRD to use
+	ChannelTemplate ChannelTemplateSpec `json:"channelTemplate"`
```

Comment:
Don't we want each step to use an existing channel or create their own channel? I'm afraid this design implies that a channel is created per step.

Reply:
Sorry, use which existing channel? If you have channels you're already created, seems like pipeline does not really add that much value since you can just create your subscriptions just as easily? I was reading through the initial proposal that we talked about a while ago, and there was no mention of reusing channels, rather than the pipeline would create the channels and subscriptions necessary to wire things together, so the user would simply give a list of functions that get executed in order.

Reply:
yes make sense, now that I understand better what can be done with handlers...